service/src/lib.rs

```diff
@@ -39,59 +35,50 @@ pub trait Factory: Send + Sync {
-///
-/// Please refer to `shuttle-examples/custom-resource` for examples of how to create custom resource. For more advanced provisioning
-/// of custom resources, please [get in touch](https://discord.gg/shuttle) and detail your use case. We'll be interested to see what you
-/// want to provision and how to do it on your behalf on the fly.
-///
-/// ```
+/// Allows implementing plugins for the Shuttle main function.
 #[async_trait]
-pub trait ResourceBuilder<T> {
-    /// The type of resource this creates
-    const TYPE: Type;
+pub trait ResourceBuilder {
+    /// The type of resource this plugin creates.
+    const TYPE: resource::Type;
 
-    /// The internal config being constructed by this builder. This will be used to find cached [Self::Output].
-    type Config: Serialize;
+    /// The input config to this resource.
+    type Config: Default + Serialize;
 
-    /// The output type used to build this resource later
+    /// The output from requesting this resource.
+    /// A cached copy of this will be used if the same [`ResourceBuilder::Config`] is found for this [`ResourceBuilder::TYPE`].
     type Output: Serialize + DeserializeOwned;
 
-    /// Create a new instance of this resource builder
-    fn new() -> Self;
-
-    /// Get the internal config state of the builder
+    /// Get the config of this plugin after it has been built from its macro arguments with the builder pattern.
     ///
-    /// If the exact same config was returned by a previous deployment that used this resource, then [Self::output()]
+    /// If the exact same config was returned by a previous deployment that used this resource, then [`ResourceBuilder::output`]
     /// will not be called to get the builder output again. Rather the output state of the previous deployment
-    /// will be passed to [Self::build()].
+    /// will be passed to [`ResourceBuilder::build`].
     fn config(&self) -> &Self::Config;
 
-    /// Get the config output of this builder
+    /// Construct this resource with the help of metadata and by calling provisioner methods in the [`Factory`].
     ///
     /// This method is where the actual resource provisioning should take place and is expected to take the longest. It
     /// can at times even take minutes. That is why the output of this method is cached and calling this method can be
-    /// skipped as explained in [Self::config()].
+    /// skipped as explained in [`ResourceBuilder::config`].
+    ///
+    /// The output from this function is passed to [`IntoResource::init`].
     async fn output(self, factory: &mut dyn Factory) -> Result<Self::Output, crate::Error>;
+}
 
-    /// Build this resource from its config output
-    async fn build(build_data: &Self::Output) -> Result<T, crate::Error>;
+/// Implement this on an [`ResourceBuilder::Output`] type to turn the
+/// base resource into the end type exposed to the Shuttle main function.
+#[async_trait]
+pub trait IntoResource<R>: Serialize + DeserializeOwned {
```

Comment:
Can we drop the `Serialize` and `DeserializeOwned`? I'm thinking some final outputs might not always have these implemented. Also, don't see the need in having them implemented.

Reply:
`R` is the final output with no bounds. `Serialize + DeserializeOwned` are required for what we are converting from, which will always be a `ResourceBuilder::Output`.

Reply:
Got it, thanks! :pray: